nemo/collections/asr/modules/conformer_encoder.py

```diff
@@ -0,0 +1,303 @@
+            Defaults to 0.1.
+        dropout_emb (float): the dropout rate used for the positional embeddings
+            Defaults to 0.1.
+        dropout_att (float): the dropout rate used for the attention layer
+            Defaults to 0.0.
+    """
+
+    def _prepare_for_export(self):
+        Exportable._prepare_for_export(self)
+
+    def input_example(self):
+        """
+        Generates input examples for tracing etc.
+        Returns:
+            A tuple of input examples.
+        """
+        input_example = torch.randn(16, self._feat_in, 256).to(next(self.parameters()).device)
+        return tuple([input_example])
+
+    @property
+    def input_types(self):
+        """Returns definitions of module input ports.
+        """
+        return OrderedDict(
+            {
+                "audio_signal": NeuralType(('B', 'D', 'T'), SpectrogramType()),
+                "length": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    @property
+    def output_types(self):
+        """Returns definitions of module output ports.
+        """
+        return OrderedDict(
+            {
+                "outputs": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation()),
+                "encoded_lengths": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    def __init__(
+        self,
+        feat_in,
+        n_layers,
+        d_model,
+        feat_out=-1,
+        subsampling='vggnet',
+        subsampling_factor=4,
+        subsampling_conv_channels=64,
+        ff_expansion_factor=4,
+        self_attention_model='rel_pos',
+        n_heads=4,
+        xscaling=True,
+        conv_kernel_size=31,
+        dropout=0.1,
+        dropout_emb=0.1,
+        dropout_att=0.0,
+    ):
+        super().__init__()
+
+        d_ff = d_model * ff_expansion_factor
+        self.d_model = d_model
+        self.scale = math.sqrt(self.d_model)
+
+        if xscaling:
+            self.xscale = math.sqrt(d_model)
+        else:
+            self.xscale = None
+
+        if subsampling:
+            self.pre_encode = ConvSubsampling(
+                subsampling=subsampling,
+                subsampling_factor=subsampling_factor,
+                feat_in=feat_in,
+                feat_out=d_model,
+                conv_channels=subsampling_conv_channels,
+                activation=nn.ReLU(),
+            )
+            self._feat_out = d_model
+        else:
+            self._feat_out = d_model
+            self.pre_encode = nn.Linear(feat_in, d_model)
+
+        if self_attention_model == "rel_pos":
+            self.pos_enc = RelPositionalEncoding(
+                d_model=d_model, dropout_rate=dropout, dropout_emb_rate=dropout_emb, xscale=self.xscale
+            )
+        elif self_attention_model == "abs_pos":
+            self.pos_enc = PositionalEncoding(
+                d_model=d_model, dropout_rate=dropout, max_len=6000, reverse=False, xscale=self.xscale
```

Comment:
Please consider allowing the user to pass `max_len`, and maybe also `reverse` instead of hard coding it.

Reply:
+ Added a parameter to control the max_len.
+ reverse should be always False when PositionalEncoding is created directly. It is just True when RelPositionalEncoding calls its init. User should not have control on it.